server/plugin.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"github.com/pkg/errors"
+)
+
+const (
+	DelayInSecons = 6
+)
+
+type Plugin struct {
+	plugin.MattermostPlugin
+	configurationLock sync.RWMutex
+	configuration     *configuration
+	Store             *DBStore
+	botID             string
+}
+
+func (p *Plugin) OnActivate() error {
+	config := p.API.GetUnsanitizedConfig()
+	store, err := NewDBStore(*config.SqlSettings.DriverName, *config.SqlSettings.DataSource)
+	if err != nil {
+		return err
+	}
+	p.Store = store
+	if err = p.API.RegisterCommand(getCommand()); err != nil {
+		return err
+	}
+
+	recommendBot := &model.Bot{
+		Username:    "recommend-bot",
+		DisplayName: "Recommend Bot",
+		Description: "A bot account created by the com.github.jespino.recommend plugin",
+	}
+
+	options := []plugin.EnsureBotOption{
+		plugin.ProfileImagePath("assets/icon.png"),
+	}
+
+	botID, ensureBotError := p.Helpers.EnsureBot(recommendBot, options...)
+	if ensureBotError != nil {
+		return errors.Wrap(ensureBotError, "failed to ensure recommend bot user.")
+	}
+
+	p.botID = botID
+
+	return nil
+}
+
+func (p *Plugin) OnDeactivate() error {
+	p.Store.Close()
+	return nil
+}
+
+func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
+	router := mux.NewRouter()
+	router.ServeHTTP(w, r)
+}
+
+func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.ChannelMember, user *model.User) {
+	if !p.getConfiguration().RecommendOnJoinChannel {
+		return
+	}
+	time.Sleep(DelayInSecons * time.Second)
```

Comment:
why do we need a delay here?

Reply:
Because we have to wait until you have opened properly the channel, before we sent you the ephemeral message, if not, is going to be lost.